File: DsAlgo__450/Heap/Minimum Cost of ropes.cpp
// using minheap(minimum element on top of heap)
// https://practice.geeksforgeeks.org/problems/minimum-cost-of-ropes-1587115620/1#

//logic = insr all element in minheap then pop a,b element and sum them to ans, then push the new rope length(a+b) to minheap

long long minCost(long long arr[], long long n) {
    long long ans=0;
    //minheap
    priority_queue <long long, vector<long long>, greater<long long> > mnhp(arr,arr+n);

    // for(long long i=0; i<n; i++) {
    //     mnhp.push(arr[i]);
    // }

    while(mnhp.size() != 1) {
        long long a= mnhp.top();
        mnhp.pop();
        long long b= mnhp.top();
        mnhp.pop();

        ans+=(a+b);

        mnhp.push(a + b);
    }

    return ans;
}

File: DsAlgo__450/Heap/Minimum Cost of ropes.java
// using minheap(minimum element on top of heap)
// https://practice.geeksforgeeks.org/problems/minimum-cost-of-ropes-1587115620/1#

//logic = insr all element in minheap then pop a,b element and sum them to ans, then push the new rope length(a+b) to minheap
class Solution
{
    //Function to return the minimum cost of connecting the ropes.
    long minCost(long arr[], int n) 
    {
        PriorityQueue<Long> pq = new PriorityQueue<Long>();
        long  ans=0;
        for(long i:arr){
            pq.add(i);
        }

        while(pq.size() != 1){
            long a=pq.poll();
            long b=pq.poll();

            ans+=a+b;
            pq.add(a+b);
        }

        return ans;
    }
}